src/rpc/pd/timestamp.rs

```diff
@@ -2,227 +2,193 @@
+            let mut pending_requests = pending_requests.borrow_mut();
+
+            // Wake up the sending future blocked by too many pending requests as we are consuming
+            // some of them here.
+            if pending_requests.len() == MAX_PENDING_COUNT {
+                sending_future_waker.wake();
             }
-            Task::Init => PdReactor::init(reactor, pd_client, handle),
+
+            allocate_timestamps(&resp, &mut pending_requests)?;
         }
-    }
-}
+        Err(Error::internal_error("TSO stream terminated"))
+    };
 
-/// A special-purpose event loop for asynchronously requesting timestamps. This is
-/// more complex than just sending a request since requests are batched on the client.
-pub(super) struct PdReactor {
-    // These fields are for communicating internally within the reactor to initialize
-    // it and send communicate between threads.
-    task_tx: Option<UnboundedSender<Task>>,
-    tso_tx: Sender<pdpb::TsoRequest>,
-    tso_rx: Option<Receiver<pdpb::TsoRequest>>,
-
-    // These fields are for communicating with the PD cluster to get batches of timestamps.
-    handle: Option<JoinHandle<()>>,
-    tso_pending: Option<Vec<TsoChannel>>,
-    tso_buffer: Option<Vec<TsoChannel>>,
-    tso_batch: Vec<TsoChannel>,
+    let (send_res, recv_res): (_, Result<()>) = join!(send_requests, receive_and_handle_responses);
+    error!("TSO send error: {:?}", send_res);
+    error!("TSO receive error: {:?}", recv_res);
 }
 
-impl Drop for PdReactor {
-    fn drop(&mut self) {
-        if let Some(handle) = self.handle.take() {
-            handle.join().unwrap();
-        }
-    }
+struct TsoRequestStream<'a> {
+    cluster_id: u64,
+    request_rx: Pin<&'a mut mpsc::Receiver<oneshot::Sender<Timestamp>>>,
+    pending_requests: Rc<RefCell<VecDeque<TimestampRequest>>>,
+    self_waker: Rc<AtomicWaker>,
 }
 
-impl PdReactor {
-    pub(super) fn new() -> Self {
-        let (tso_tx, tso_rx) = mpsc::channel(1);
-        PdReactor {
-            task_tx: None,
-            tso_tx,
-            tso_rx: Some(tso_rx),
-            handle: None,
-            tso_buffer: Some(Vec::with_capacity(8)),
-            tso_batch: Vec::with_capacity(8),
-            tso_pending: None,
+impl<'a> Stream for TsoRequestStream<'a> {
+    type Item = (TsoRequest, WriteFlags);
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let pending_requests = self.pending_requests.clone();
+        let mut pending_requests = pending_requests.borrow_mut();
+        let mut count = 0;
+        while count < MAX_BATCH_SIZE && pending_requests.len() < MAX_PENDING_COUNT {
+            match self.request_rx.as_mut().poll_next(cx) {
+                Poll::Ready(Some(sender)) => {
+                    pending_requests.push_back(sender);
+                    count += 1;
+                }
+                Poll::Ready(None) => return Poll::Ready(None),
```

Comment:
If count > 0 now, may we drop the received requests?

Reply:
Yes and it is expected. As we don't close the channel manually, the only possibility we go into this branch is that all senders are dropped. And now, this only happens when a reconnection occurs and the old `TimestampOracle` in 
the `RetryClient` is replaced with a new one. Then, the old channel to the PD server is no longer usable. We can simply drop them here and finally the receiver will get an error.